### ISSUE TITLE
Unable to Store Chinese Character '𧗽' in MySQL with utf8mb4 Charset

### ISSUE THREAD

**Issue by CasperHK** [opis/database#102]
I'm facing an issue while trying to store the Chinese character '𧗽' (U+275FD) in a MySQL database table that uses the utf8mb4 character set. I have verified that both the database and the specific table are set to utf8mb4, but I'm still receiving an error when attempting to insert this character.

Here are the details of my setup:

| Attribute | Value  | 
|---|---|
| MySQL Version | 8.4.2 | 
| PHP Version | 8.3.12 |
| Table Charset | utf8mb4 | 
| Table Collation | utf8mb4_unicode_ci |
| Column Charset | utf8mb4 |

The character 𧗽 (U+26C7D) is a CJK (Chinese, Japanese, Korean) character that falls outside the range of the utf8 character set but is supported by utf8mb4. This is likely causing the collation issue you encountered.

## Error Message
```console
General error: 1366 Incorrect string value: 'U+275FD' for column 'chinese' at row 1 [ INSERT INTO `Staff` (`staff_no`, `name`, `chinese`, `TYPE`, `email`, `sex`) VALUES ('12345', 'HANG HANG HANG', '𧗽', 'D', 'hhh_03@example.com.hk', 'M') ]  in /var/www/html/evs/vendor/opis/database/src/Connection.php:567
Stack trace:
#0 /var/www/html/evs/vendor/opis/database/src/Connection.php(426): Opis\Database\Connection->execute()
#1 /var/www/html/evs/vendor/opis/database/src/SQL/Insert.php(47): Opis\Database\Connection->command()
#2 /var/www/html/evs/cron/night-job.php(77): Opis\Database\SQL\Insert->into()
#3 {main}
  thrown in /var/www/html/evs/vendor/opis/database/src/Connection.php on line 567
```

I would appreciate any insights into why this character isn't being stored correctly and any steps I can take to resolve the issue. Thank you!

**Comment by CasperHK** [comment]
I have tested saving the same Chinese character into MySQL using DBeaver and MS-SQL, both works.

The DBeaver is able to update the Chinese attribute by simply:ble to update the Chinese attribute by simply:
```sql
--  Auto-generated SQL script. Actual values for binary/complex data types may differ - what you see is the default string representation of values.
UPDATE voting_system.Staff
	SET chinese='𧗽𧗽𧗽'
	WHERE staff_no='12345';
```
Without any error.

I also tested inserting '𧗽𧗽𧗽' in an online PHP SQL Tester, and it works.
https://phpize.online/sql/mysql80/25aeb2883aa147e84760bba05a624d52/php/php82/e765196b03fa5c6f303320e6ffbf870e/

**Comment by sorinsarca** [comment]
Try using an init command https://opis.io/database/4.x/connections.html#init-commands

**Comment by CasperHK** [comment]
It seems works but...
```php
Fatal error: Uncaught PDOException: SQLSTATE[IMSSP]: An invalid keyword 'host' was specified in the DSN string. in /var/www/html/evs/vendor/opis/database/src/Connection.php:290
Stack trace:
#0 /var/www/html/evs/vendor/opis/database/src/Connection.php(290): PDO->__construct()
#1 /var/www/html/evs/vendor/opis/database/src/Connection.php(262): Opis\Database\Connection->getPDO()
#2 /var/www/html/evs/vendor/opis/database/src/Connection.php(308): Opis\Database\Connection->getDriver()
#3 /var/www/html/evs/vendor/opis/database/src/SQL/Select.php(48): Opis\Database\Connection->getCompiler()
#4 /var/www/html/evs/vendor/opis/database/src/SQL/Query.php(163): Opis\Database\SQL\Select->select()
#5 /var/www/html/evs/cron/night-job.php(49): Opis\Database\SQL\Query->select()
#6 {main}
  thrown in /var/www/html/evs/vendor/opis/database/src/Connection.php on line 290
```

**Comment by sorinsarca** [comment]
I cannot help you without providing the code where you make the connection.

**Comment by CasperHK** [comment]
I connect the database by calling the function defined below.
```php
use Opis\Database\Database;
use Opis\Database\Connection;

function createConnection($conn_infos) {
  $connection = new Connection(
    $conn_infos["DB_TYPE"].':host='.$conn_infos["HOST"].';dbname='.$conn_infos["DB_NAME"],
    $conn_infos["USERNAME"],
    $conn_infos["PASSWORD"],
  );
  $connection->initCommand('SET NAMES utf8mb4');
  return $connection;
}

function createSQLSVConnection($conn_infos) {
  $connection = new Connection(
    $conn_infos["DB_TYPE"].':Server='.$conn_infos["HOST"].';Database='.$conn_infos["DB_NAME"].';TrustServerCertificate=true',
    $conn_infos["USERNAME"],
    $conn_infos["PASSWORD"]
  );
  $connection->initCommand('SET NAMES utf8mb4');
  return $connection;
}
```

## Error
```
Fatal error: Uncaught Error: Call to undefined method mysqli::initCommand() in C:\xampp\htdocs\EVS\com\db_conn.php:28 Stack trace: #0 C:\xampp\htdocs\EVS\router.php(27): createConnection2(Array) #1 {main} thrown in C:\xampp\htdocs\EVS\com\db_conn.php on line 28
```

**Comment by CasperHK** [comment]
```php
Fatal error: Uncaught PDOException: SQLSTATE[HY000]: General error: 3988 Conversion from collation utf8mb4_0900_ai_ci into utf8_general_ci impossible for parameter [ UPDATE `Election` SET `title_en` = 'stdhwsht𧗽', `title_tc` = 'ethrwsth', `period_from` = 1729134600, `period_to` = 1730430600, `descr` = 'etrhwhtr', `last_upd_by` = 1, `last_upd_unix` = 1729566695 WHERE `id` = '166' ] in C:\xampp\htdocs\EVS\vendor\opis\database\src\Connection.php:573 Stack trace: #0 C:\xampp\htdocs\EVS\vendor\opis\database\src\Connection.php(441): Opis\Database\Connection->execute(Array) #1 C:\xampp\htdocs\EVS\vendor\opis\database\src\SQL\Update.php(100): Opis\Database\Connection->count('UPDATE `Electio...', Array) #2 C:\xampp\htdocs\EVS\handlers\api\edit-election.php(24): Opis\Database\SQL\Update->set(Array) #3 [internal function]: {closure}('166') #4 C:\xampp\htdocs\EVS\vendor\bramus\router\src\Bramus\Router\Router.php(458): call_user_func_array(Object(Closure), Array) #5 C:\xampp\htdocs\EVS\vendor\bramus\router\src\Bramus\Router\Router.php(440): Bramus\Router\Router->invoke(Object(Closure), Array) #6 C:\xampp\htdocs\EVS\vendor\bramus\router\src\Bramus\Router\Router.php(287): Bramus\Router\Router->handle(Array, true) #7 C:\xampp\htdocs\EVS\router.php(244): Bramus\Router\Router->run() #8 {main} thrown in C:\xampp\htdocs\EVS\vendor\opis\database\src\Connection.php on line 573
```

```sql
Error synchronizing data with database

Reason:
SQL Error [1366] [HY000]: Incorrect string value: '\xF0\xA7\x97\xBD' for column 'title_tc' at row 1
```

**Comment by CasperHK** [comment]
> I cannot help you without providing the code where you make the connection.

Here is the error code:
```php
$r = $db_mysql->from('Staff')->delete();

$rows = $db_mssql->from('ELECTION_STAFF')
                    ->select([
                        "staff_no" => "STAFF_NO",
                        "name"     => "NAME",
                        "chinese"  => "CHINESE",
                        "RANK"   => "RANK",
                        "TYPE"   => "TYPE",
                        "inst_t" => "INST_T",
                        "inst_x" => "INST_X",
                        "section_t"  => "SECTION_T",
                        "section_x"  => "SECTION_X",
                        "exact_inst" => "EXACT_INST",
                        "inst_r" => "INST_R",
                        "regoin" => "REGION",
                        "email" => "EMAIL",
                        "sex" => "SEX"
                    ])->all();

foreach ($rows as $record) {
    $result = $db_mysql->insert([
        "staff_no" => $record->STAFF_NO,
        "name"     => $record->NAME,
        "chinese"  => $record->CHINESE,
        "RANK"     => $record->RANK,
        "TYPE"     => $record->TYPE,
        "inst_t"   => $record->INST_T,
        "inst_x"   => $record->INST_X,
        "section_t"  => $record->SECTION_T,
        "section_x"  => $record->SECTION_X,
        "exact_inst" => $record->EXACT_INST,
        "inst_r" => $record->INST_R,
        "region" => $record->REGION,
        "email"  => $record->EMAIL,
        "sex" => $record->SEX,
    ])
    ->into('Staff');    // Error Happen
}
```

The PHP code performs the following operations:

1. **Delete Existing Records**: It deletes all records from the Staff table in a MySQL database.
2. **Fetch Records from MSSQL**: It retrieves a set of staff records from the ELECTION_STAFF table in a Microsoft SQL Server database. The selected fields include STAFF_NO, NAME, CHINESE, RANK, TYPE, INST_T, INST_X, SECTION_T, SECTION_X, EXACT_INST, INST_R, REGION, EMAIL, and SEX.
3. **Insert New Records**: For each record fetched from the MSSQL database, it inserts the corresponding data into the Staff table in the MySQL database.

In summary, the code clears the MySQL Staff table and populates it with the latest staff information from an MS-SQL source.

## Connection Method
```php
function createConnection(): Connection {
    $connection = new Connection(
        $this->connInfos["DB_TYPE"] . ':host=' . $this->connInfos["HOST"] . ';dbname=' . $this->connInfos["DB_NAME"],
        $this->connInfos["USERNAME"],
        $this->connInfos["PASSWORD"]
    );
    $connection->initCommand('SET NAMES utf8mb4');
    return $connection;
}
```

**Comment by CasperHK** [comment]
Problem finally resolved by changing all the tables and columns to `utf8mb4_bin`.
